keps/sig-node/20200310-ensure-secret-pulled-images.md

```diff
@@ -0,0 +1,169 @@
+`ensureSecretPulledImages` to the pod spec. The default to false
+means that if a first pod results in an image pulled with imagePullSecrets a
+second pod would have to be using always pull to ensure rights to use the
+previously pulled image. When set to true always pull would not be required,
+instead kublet will check if the image was pulled with an image pull secret and
+if so would force a pull of the image to ensure the image pulled with the
+secret is not used by another pod unless that pod also has the proper auth.
+
+## Motivation
+
+There have been customer requests for improving upon kubernetes ability to
+secure images pulled with auth. on a node. Issue
+[#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
+for a while.
+
+To secure images one currently needs to inject `AllwaysPullImages` into pod
+specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+
+### Goals
+
+Add a flag processed by `kubelet` for `ensureSecretPulledImages` (or something
+similarly named) that, if true, would force `kubelet` to attempt to pull every
+image that was pulled with image pulled secret based authentication, regardless
+of the container image's pull policy.
+
+Optimize to only force re-authentication for a pod when the secret used to pull
+the container image is not present.
+
+### Non-Goals
+
+Out of scope for this KEP is an image caching policy that would direct container
+runtimes through the CRI wrt. how they should treat the caching of images on a
+node. Such as store for public use but only if encrypted. Or Store for private
+use unencrypted...
+
+## Proposal
+
+When `ensureSecretPulledImages` is set, `kublet` will check keep a list of
+container images that required authentication. `kublet` will ensure any image
+in the list is always pulled thus enforcing authentication / re-authentication
+with the exception of pods with secrets containing an auth that has been
+authenticated.
+
+### User Stories
+wip
+
+### Risks and Mitigations
+
+With the default being false, devops engineers may not know to set the flag to
+true in new/old pod specs that are using secrets for pull authentication with
+registries.
+
+A mitigation would be an admission plugin to inject `ensureSecretPulledImages.`
+
+Images authentications with a registry may expire. To mitigate expirations a
```

Comment:
How will we determine what these timeouts should be?

Similarly, confirming that if the secrets which initially allowed the image to be downloaded are revoked, the pod will still be allowed to use the image until the timeout is reached?

Reply:
Yes, this is a problem.. Without having more data from the registry / container runtime we lack the ability to know things like how long is this user authorized to use this image. We can have configurable defaults to force retrying the authentication but this is a case where we (kublet) are trying to drive from the back seat. We could start with 1day or whatever the industry standard is, and let the DevOps engineers make a decision on performance / security / privacy etc.

Reply:
An optional timeout setting is a stretch goal and could possibly should be resolved at the container runtime cache level. Is not the primary goal of this KEP.